File: packages/agent/src/sub-agent-tool.ts
import type {
  ITool,
  ILlmProvider,
  ToolDescription,
  ToolResult,
  JsonObject,
  AgentLogger,
} from '@cli-agent/core';
import { Registry, RunContext, createChildLogger } from '@cli-agent/core';
import { AgentLoop } from './agent-loop.js';
import type { SystemPromptBuilder } from './agent-loop.js';
import type { PermissionHandler } from './permission.js';

/** Configuration for creating a sub-agent tool. */
export interface SubAgentToolConfig {
  /** Display name for this sub-agent tool */
  readonly name: string;
  /** Description shown to the parent agent */
  readonly description: string;
  /** LLM provider the sub-agent will use */
  readonly provider: ILlmProvider;
  /** Tools available to the sub-agent */
  readonly toolRegistry: Registry<ITool>;
  /** System prompt or dynamic builder for the sub-agent */
  readonly systemPrompt?: string;
  readonly systemPromptBuilder?: SystemPromptBuilder;
  /** Max iterations for the sub-agent (default: 25) */
  readonly maxIterations?: number;
  /** Permission handler for the sub-agent's tools */
  readonly permissionHandler?: PermissionHandler;
}

const DEFAULT_MAX_ITERATIONS = 25;

/**
 * Wraps an AgentLoop as an ITool so a parent agent can delegate
 * sub-tasks to a child agent via tool calls.
 *
 * The parent sends a "task" parameter; the sub-agent runs autonomously
 * and returns the final result as tool output.
 */
export class SubAgentTool implements ITool {
  readonly name: string;
  readonly requiresPermission = false;

  private readonly config: SubAgentToolConfig;
  private readonly logger: AgentLogger;

  constructor(config: SubAgentToolConfig) {
    this.name = config.name;
    this.config = config;
    this.logger = createChildLogger(`sub-agent:${config.name}`);
  }

  describe(): ToolDescription {
    return {
      name: this.name,
      description: this.config.description,
      parameters: [
        {
          name: 'task',
          type: 'string',
          description: 'The task to delegate to the sub-agent',
          required: true,
        },
      ],
    };
  }

  async execute(params: JsonObject, context: RunContext): Promise<ToolResult> {
    const task = params['task'];
    if (typeof task !== 'string' || task.trim().length === 0) {
      return { success: false, output: '', error: 'Missing or empty "task" parameter' };
    }

    this.logger.info({ task: task.slice(0, 200) }, 'Sub-agent starting');

    const childLoop = new AgentLoop({
      provider: this.config.provider,
      toolRegistry: this.config.toolRegistry,
      config: {
        provider: {
          providerId: this.config.provider.providerId,
          model: 'sub-agent',
          auth: { type: 'api-key' as const, apiKey: '' },
          maxTokens: 4096,
          temperature: 0.7,
        },
        maxIterations: this.config.maxIterations ?? DEFAULT_MAX_ITERATIONS,
        workingDirectory: context.workingDirectory,
        systemPrompt: this.config.systemPrompt,
      },
      permissionHandler: this.config.permissionHandler,
      systemPromptBuilder: this.config.systemPromptBuilder,
    });

    // Propagate abort from parent context to child via AbortSignal
    const onAbort = () => childLoop.abort('Parent aborted');
    context.signal.addEventListener('abort', onAbort, { once: true });

    try {
      const result = await childLoop.run(task);

      this.logger.info(
        { iterations: result.iterations, aborted: result.aborted },
        'Sub-agent completed',
      );

      if (result.aborted) {
        return { success: false, output: result.content, error: 'Sub-agent was aborted' };
      }

      return { success: true, output: result.content };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      this.logger.warn({ error: message }, 'Sub-agent failed');
      return { success: false, output: '', error: `Sub-agent error: ${message}` };
    } finally {
      context.signal.removeEventListener('abort', onAbort);
    }
  }
}

File: sdk/python/chameleon/sub_agent_tool.py
"""
SubAgentTool — Wraps an AgentLoop as an ITool.
Python port of @cli-agent/agent SubAgentTool.

The parent agent calls this like any other tool:
  news_researcher(task="오늘 한국 뉴스 검색해줘")

Internally it spawns a child AgentLoop that runs autonomously.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass

from .types import (
    ITool,
    ILlmProvider,
    ToolDescription,
    ToolParameter,
    ToolResult,
    RunContext,
    JsonObject,
)
from .registry import Registry
from .agent_loop import AgentLoop, AgentLoopOptions

logger = logging.getLogger(__name__)

DEFAULT_SUB_MAX_ITERATIONS = 10


@dataclass
class SubAgentToolConfig:
    """Configuration for creating a sub-agent tool."""

    name: str
    description: str
    provider: ILlmProvider
    tool_registry: Registry[ITool]
    system_prompt: str = ""
    max_iterations: int = DEFAULT_SUB_MAX_ITERATIONS


class SubAgentTool(ITool):
    """
    An ITool that delegates work to a child AgentLoop.

    ```python
    tool = SubAgentTool(SubAgentToolConfig(
        name="news_researcher",
        description="뉴스 전문 리서치",
        provider=provider,
        tool_registry=search_tools,
        system_prompt="당신은 뉴스 전문 리서치 에이전트입니다.",
    ))
    registry.register("news_researcher", tool)
    ```
    """

    def __init__(self, config: SubAgentToolConfig) -> None:
        self._config = config

    @property
    def name(self) -> str:
        return self._config.name

    def describe(self) -> ToolDescription:
        return ToolDescription(
            name=self._config.name,
            description=self._config.description,
            parameters=[
                ToolParameter(
                    name="task",
                    type="string",
                    description="The task to delegate to the sub-agent",
                    required=True,
                ),
            ],
        )

    async def execute(self, params: JsonObject, context: RunContext) -> ToolResult:
        task = params.get("task", "")
        if not isinstance(task, str) or not task.strip():
            return ToolResult(
                success=False,
                output="",
                error='Missing or empty "task" parameter',
            )

        logger.info("Sub-agent '%s' starting: %s", self._config.name, task[:200])

        child_loop = AgentLoop(
            AgentLoopOptions(
                provider=self._config.provider,
                tool_registry=self._config.tool_registry,
                system_prompt=self._config.system_prompt,
                max_iterations=self._config.max_iterations,
                working_directory=context.working_directory,
            )
        )

        # Propagate abort
        if context.aborted:
            return ToolResult(
                success=False,
                output="",
                error="Parent context already aborted",
            )

        try:
            result = await child_loop.run(task)

            logger.info(
                "Sub-agent '%s' completed: iterations=%d, aborted=%s",
                self._config.name,
                result.iterations,
                result.aborted,
            )

            if result.aborted:
                return ToolResult(
                    success=False,
                    output=result.content,
                    error="Sub-agent was aborted",
                )

            return ToolResult(success=True, output=result.content)

        except Exception as e:
            logger.exception("Sub-agent '%s' failed", self._config.name)
            return ToolResult(
                success=False,
                output="",
                error=f"Sub-agent error: {e}",
            )
